### PR TITLE
Ema/fix mutating forms

### PR DESCRIPTION
**Pull request by GioSensation**
**Reviewer:** @shakyShane 
**Asana:** https://app.asana.com/0/0/1205015898566987/f

## Description
Observe form mutations and re-scan the form when input fields are removed. This won't catch all possible forms of mutations, but it's a good first step with reduced overhead.

## Steps to test
Test case added. You can test this live on these sites:
- https://costplusdrugs.com
- https://scite.ai/
- https://www.espn.com/

### CODE REVIEW

File: src/Form/Form.js [hunk 1]
```diff
@@ -45,9 +43,6 @@ class Form {
         this.form = form
         this.matching = matching || createMatching()
         this.formAnalyzer = new FormAnalyzer(form, input, matching)
-        this.isLogin = this.formAnalyzer.isLogin
-        this.isSignup = this.formAnalyzer.isSignup
-        this.isHybrid = this.formAnalyzer.isHybrid
```
Comment (GioSensation) [inline src/Form/Form.js:50]: These now are all getters, so if we change the `formAnalyzer` results, the changes are reflected here.

File: src/Form/Form.js [hunk 2]
```diff
@@ -77,6 +72,23 @@ class Form {
             }
         })
 
+        this.mutObsConfig = { childList: true, subtree: true }
+        this.mutObs = new MutationObserver(
+            (records) => {
+                const anythingRemoved = records.some(record => record.removedNodes.length > 0)
+                if (anythingRemoved) {
```
Comment (GioSensation) [inline src/Form/Form.js:79]: By running only when elements are removed we reduce the overhead.

File: src/Form/FormAnalyzer.js [hunk 3]
```diff
@@ -9,6 +9,8 @@ const signupRegex = new RegExp(
 )
 const conservativeSignupRegex = new RegExp(/sign.?up|join|register|enroll|newsletter|subscri(be|ption)|settings|preferences|profile|update/i)
 const strictSignupRegex = new RegExp(/sign.?up|join|register|(create|new).+account|enroll|settings|preferences|profile|update/i)
+const resetPasswordLink = new RegExp(/(forgot(ten)?|reset|don't remember) (your )?password|password forgotten/i)
+const loginProvidersRegex = new RegExp(/ with /i)
```
Comment (GioSensation) [inline src/Form/FormAnalyzer.js:13]: Changes here are unrelated to the mutating forms. Just improving accuracy and splitting these two regexes so we can use them independently.

File: src/Form/Form.js [hunk 4]
```diff
@@ -77,6 +77,30 @@ class Form {
...
+                        window.requestIdleCallback(() => {
+                            this.formAnalyzer = new FormAnalyzer(this.form, input, this.matching)
+                            this.recategorizeAllInputs()
+                        })
+
+                        this.mutObsCount++
+                        // If the form mutates too much, disconnect to avoid performance issues
+                        if (this.mutObsCount >= MAX_FORM_MUT_OBS_COUNT) {
+                            this.mutObs.disconnect()
+                        }
```
Comment (GioSensation) [inline src/Form/Form.js:98]: Added this bit to disconnect the observer when form mutations force too many re-scans. This prevents infinite loops and other performance issues.

File: src/Form/Form.js [hunk 5]
```diff
@@ -323,8 +362,14 @@ class Form {
         if (this.form.matches(selector)) {
             this.addInput(this.form)
         } else {
-            this.form.querySelectorAll(selector).forEach(input => this.addInput(input))
+            const foundInputs = this.form.querySelectorAll(selector)
```
Comment (GioSensation) [inline src/Form/Form.js:365]: This is a further performance safeguard. If the form contains too many inputs, we stop adding more.

File: src/Form/Form.js [hunk 6]
```diff
@@ -378,11 +423,25 @@ class Form {
     }
 
     addInput (input) {
+        if (this.inputs.all.has(input)) return this
+
+        // If the form has too many inputs, destroy everything to avoid performance issues
+        if (this.inputs.all.size > MAX_INPUTS_PER_FORM) {
```
Comment (GioSensation) [inline src/Form/Form.js:429]: And here we `destroy()` the form, removing all listeners and icons, if the form grows over a certain threshold.

File: src/Form/Form.js [hunk 7]
```diff
@@ -378,11 +423,25 @@ class Form {
...
+        if (this.initialScanComplete) {
+            this.formAnalyzer = new FormAnalyzer(this.form, input, this.matching)
+            this.recategorizeAllInputs()
+            return this
+        }
+
         // Nothing to do with 1-character fields
         if (input.maxLength === 1) return this
 
-        if (this.inputs.all.has(input)) return this
```
Comment (GioSensation) [inline src/Form/Form.js:384]: Moved to the top to avoid unnecessary work.

File: src/Scanner.js [hunk 8]
```diff
@@ -32,7 +42,9 @@ const defaultScannerOptions = {
     // How many inputs is too many on the page. If we detect that there's above
     // this maximum, then we don't scan the page. This will prevent slowdowns on
     // large pages which are unlikely to require autofill anyway.
-    maxInputsOnPage: 100
+    maxInputsPerPage: MAX_INPUTS_PER_PAGE,
+    maxFormsPerPage: MAX_FORMS_PER_PAGE,
+    maxInputsPerForm: MAX_INPUTS_PER_FORM
```
Comment (GioSensation) [inline src/Scanner.js:47]: Centralizing these defaults in the constants.js file.

File: src/UI/HTMLTooltip.js [hunk 9]
```diff
@@ -27,14 +27,14 @@ export const defaultOptions = {
...
+            transform: translate(${Math.floor(left)}px, ${Math.floor(top)}px) !important;
         }
     `,
     caretPositionClass: (top, left, isAboveInput) => `
         .tooltip--email__caret {
             ${isAboveInput
-        ? `transform: translate(${left}px, ${top}px) rotate(180deg); transform-origin: 16px !important;`
-        : `transform: translate(${left}px, ${top}px) !important;`
+        ? `transform: translate(${Math.floor(left)}px, ${Math.floor(top)}px) rotate(180deg); transform-origin: 18px !important;`
+        : `transform: translate(${Math.floor(left)}px, ${Math.floor(top)}px) !important;`
```
Comment (GioSensation) [inline src/UI/HTMLTooltip.js:37]: All this stuff and other CSS changes are of course unrelated. Just squeezing in a 2px reduction in icon size requested by the design folks. The `floor` is just to avoid weird artefacts in edge cases.

File: src/Form/Form.js (hunk 4 above)
Comment (shakyShane) [reply src/Form/Form.js:98]: excellent!

File: src/Form/Form.js (hunk 6 above)
Comment (shakyShane) [reply src/Form/Form.js:429]: love it!

File: src/Form/Form.js [hunk 10]
```diff
@@ -323,8 +362,14 @@ class Form {
         if (this.form.matches(selector)) {
             this.addInput(this.form)
         } else {
-            this.form.querySelectorAll(selector).forEach(input => this.addInput(input))
+            const foundInputs = this.form.querySelectorAll(selector)
+            if (foundInputs.length < MAX_INPUTS_PER_FORM) {
+                foundInputs.forEach(input => this.addInput(input))
+            } else {
+                console.log('The form has too many inputs, bailing.')
```
Comment (shakyShane) [inline src/Form/Form.js:369]: needs wrapping in a debug check

File: src/Form/Form.js [hunk 11]
```diff
@@ -378,11 +423,25 @@ class Form {
     }
 
     addInput (input) {
+        if (this.inputs.all.has(input)) return this
+
+        // If the form has too many inputs, destroy everything to avoid performance issues
+        if (this.inputs.all.size > MAX_INPUTS_PER_FORM) {
+            console.log('The form has too many inputs, destroying.')
```
Comment (shakyShane) [inline src/Form/Form.js:430]: needs wrapping in a debug check?

File: src/Scanner.js [hunk 12]
```diff
@@ -200,7 +212,12 @@ class DefaultScanner {
                 this.forms.delete(childForm)
             }
 
-            this.forms.set(parentForm, new Form(parentForm, input, this.device, this.matching, this.shouldAutoprompt))
+            // Only add the form if below the limit of forms per page
+            if (this.forms.size < this.options.maxFormsPerPage) {
+                this.forms.set(parentForm, new Form(parentForm, input, this.device, this.matching, this.shouldAutoprompt))
+            } else {
+                console.log('The page has too many forms, stop adding them.')
```
Comment (shakyShane) [inline src/Scanner.js:219]: wrap in debug check?

**Review by shakyShane — CHANGES_REQUESTED** [review]
Looks great - I think we just need to be conservative with logging, and then this is good to go :)

**Review by shakyShane — APPROVED** [review]
